### PR TITLE
Remove ModelTrainingConfig class, refactor yaml configuration

### PR DESCRIPTION
**Pull request by mcgibbon**
We have added a new configuration infrastructure for configuring datasets and training, but have so far kept backwards compatibility with previous yaml configurations. This PR removes the previous yaml configuration infrastructure for ModelTrainingConfig used by the train and offline diags workflows, and uses the new configuration classes instead.

This is used in a workflow control example in https://github.com/VulcanClimateModeling/vcm-workflow-control/pull/110.

Open questions for reviewers:
- Right now the argo workflows pass identical data configuration to the training code for validation and to the offline diags for evaluation. Should this be decoupled into two different configuration files?
- `drop_temperature_humidity_tendencies_if_not_predicted` has been refactored to check if the model-predicted values are all set to NaN, and treat the value as not predicted if so. This is done because we no longer have a list of ML outputs, either from training config or from a model object, when generating the report in create_report. Is this OK?

Added public API:
- symbol (e.g. `vcm.my_function`) or script and optional description of changes or why they are needed
- Can group multiple related symbols on a single bullet
- Added a CLI tool `validate_batches_config` which takes in a yaml file path of a BatchesLoader and attempts to load it, raising an exception if the configuration is malformed
- Added `open_zarr` to the loaders top-level

Refactored public API:
- Refactored command-line APIs of `fv3fit.train` and `offline_ml_diags.compute_diags` to use new configuration classes
- Data configuration now must be explicitly passed to the offline ml diags workflow, it will not use output from the training step to determine this configuration
- Grid resolution is now passed explicitly to compute_diags as a command-line argument, instead of reading it from training configuration

Significant internal changes:
- Added unit tests that all registered loaders functions are imported to the top level of the package, since they are public API
- Loaders configuration classes now raise an exception on initialization if an invalid function name is specified
- Removed dependency of offline ml diags generation on training configuration file
- compute_diags now uses a `Map` with a predict function instead of PredictionMapper, which has been removed
- Refactored offline diags integration test to test create_report as well as compute_diags

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/diagnostics_utils/diagnostics_utils/utils.py
Comment (mcgibbon) [inline external/diagnostics_utils/diagnostics_utils/utils.py:42]: This variable is unused in this function, so I removed it.

File: workflows/offline_ml_diags/offline_ml_diags/_metrics.py
Comment (mcgibbon) [inline workflows/offline_ml_diags/offline_ml_diags/_metrics.py:40]: These arguments were unused.

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks for getting all the new configs working throughout the workflows and also doing some cleanup in the offline diags! My main comment is with regard to the new handling of the validation data config- previously this was an optional arg but it is now required. Unless there's a strong reason to move towards always requiring the training step to do validation, including cases where we aren't tuning hyperparams, can this be optional as it was in the past?

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py:53]: I mostly changed this to simplify the interface, since my understanding was we usually have validation data. Training codes are welcome to ignore the validation data they are given.

If we want this to be optional (which is what I'm hearing) I'll turn validation data into an optional requirement. if not passed I'll set validation batches to `[]` to avoid None checks.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:140]: Me neither!

File: workflows/offline_ml_diags/README.md
Comment (mcgibbon) [inline workflows/offline_ml_diags/README.md:62]:
```suggestion
  data_yaml             BatchesLoader configuration file.
```

**Comment by AnnaKwa** [comment]
Also could you update the description to remove some of the changes that you already merged in smaller PRs? e.g.
```
Added a CLI tool validate_batches_config which takes in a yaml file path of a BatchesLoader and attempts to load it, raising an exception if the configuration is malformed
Added open_zarr to the loaders top-level
```

File: workflows/offline_ml_diags/tests/test__ml_prediction.py
Comment (brianhenn) [inline workflows/offline_ml_diags/tests/test__ml_prediction.py:1]: Probably don't need everything in here but don't we lose test coverage by not porting these over to tests of `predict_function`?

File: workflows/offline_ml_diags/README.md
Comment (brianhenn) [inline workflows/offline_ml_diags/README.md:19]: Thanks for updating. It would be nice if we had autodocs for `offline_ml_diags` but that's out of the scope of this PR, maybe make an issue if you like

File: external/fv3fit/fv3fit/train.py
Comment (brianhenn) [inline external/fv3fit/fv3fit/train.py:17]: Nit: the documentation for this CLI ends up being a little sparse (see current: http://vulcanclimatemodeling.com/docs/fv3fit/commandline.html). Providing somewhat more verbose help strings might be worth it, for example specifying what the types of configs are that are now required for training and data.

**Review by brianhenn — COMMENTED** [review]
Thanks @mcgibbon this is a big lift and it looks like it was handled very well overall. I have a lot of minor comments. One ask is that you provide a link to an example of a trained model and an offline report generated from under the new configurations (I guess it could be the integration test, though doing it with a full-sized dataset might be better), as a check that we're happy with everything. 

To answer your question about the validation data it seems fine to pass the same dataset to the training and the offline diags for now, if we want a more elaborate train-validate-test arrangement in the future we could do that. 

Also I didn't review the argo/integration test portions much so hoping someone else can look at those.

File: external/fv3fit/fv3fit/train.py
Comment (brianhenn) [reply external/fv3fit/fv3fit/train.py:53]: I would second making it an optional argument, since I think it's more confusing to have to configure validation data for an RF routine that doesn't use it than to have to manage an optional arg.

File: external/fv3fit/tests/training/test_cli.py
Comment (nbren12) [inline external/fv3fit/tests/training/test_cli.py:128]: @mcgibbon and I discussed removing the special handling of the data_path since not all mappers are represented by one canonical data_path (e.g. the hybrid). `data_path` can be passed in `mapper_kwargs`.

This PR doesn't touch BatchesConfig so this change may be out of scope, we can address it in a follow-up PR or oepn an issue.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py:259]: I'm not sure but I can make an educated guess. Given that I did a backwards-compatible refactor of dense training to use DenseHyperparameters above, and DenseHyperparameters does not contain scaler type/kwargs, I don't think this scaler was ever used with DenseModel. We needed to do mass scaling of values for the random forest because there's no other way to manipulate the loss function, whereas for the dense model we could do mass weighting directly in the loss (I don't know if this was ever used).

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py:17]: Do you think it would be sufficient to mention the class names in the help strings here, along with the header comment "A command-line interface is available using python -m fv3fit.train. For more detail on the format of the configuration file, see configuration."?

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py:53]: @brianhenn I am planning internally to make validation data a required argument for training functions, but the random forest training function can completely ignore this argument. It will be given an empty list when validation data is not passed.

File: external/fv3fit/tests/test__config.py
Comment (mcgibbon) [reply external/fv3fit/tests/test__config.py:66]: Yes, I'll update the docs above. What I'd really like to do is remove the `additional_variables` field entirely, and instead automatically determine them based on other configuration settings (e.g. whether a derived variable is present in inputs) the way @AnnaKwa did for the offline diags, but that would be a longer task.

File: external/fv3fit/tests/training/test_cli.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_cli.py:16]: Given 48 is not used, probably not. I'll update it.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:70]: No, I thought it would but it won't. if you have cleanup you want to do in a fixture you need to put it in a `finally:` block it turns out, or it won't be triggered when the test fails.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:90]: Yeah, I could do that. I originally didn't because "training takes long" but in this case the training codes are all mocks, so it should still be fast to call main 5x as many times. I'll try to merge in the "with_derived_output_variables" test by parameterizing over its value, and having a check that only occurs if it is non-empty.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:148]: Thanks, this should be cleaned up if I can merge the two tests.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:180]: "mock_dataset" is an input to this fixture, but the timestep configuration below assumes there are 9 times present that it says to use. If the fixture for mock_dataset is updated and makes the dataset smaller than 9 times, this assert will trigger a failure.

To clear up the confusion I'll refactor mock_dataset to be a function called by this function that takes in an argument for number of timesteps. Then this assert shouldn't be necessary.

File: tests/end_to_end_integration/argo.yaml
Comment (mcgibbon) [reply tests/end_to_end_integration/argo.yaml:9]: Yes @oliverwm1 was actually the author of any changes to this file, in https://github.com/VulcanClimateModeling/fv3net/pull/1310

File: external/fv3fit/fv3fit/_shared/config.py
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/_shared/config.py:259]: The mass scaler was only compatible with the sklearn training in the past, aside from maybe existing in a dev branch somewhere.

File: workflows/offline_ml_diags/offline_ml_diags/compute_diags.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py:319]: A DerivedModel will automatically (thanks to @AnnaKwa) include the additional variables in its `input_variables`. The model itself is a black box that turns inputs into outputs, and deriving output variables using something other than "machine learning" but inside a Model is still just another part of that black box.

Comment (AnnaKwa) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py:319]: The term "additional_variables" is a bit confusing now that there's multiple scenarios in which "additional" variables are needed. (a) if needed for a derived output variable, the `DerivedModel` will handle this automatically now and add these to `DerivedModel.input_variables` (the set of inputs actually used for ML prediction are kept as `DerivedModel._base_model.input_variables`) (b) needed for offline diags computation, ex. pressure level thickeness, those are configured here.

File: workflows/offline_ml_diags/offline_ml_diags/create_report.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/offline_ml_diags/create_report.py:32]: Since these are in the same package I'll change them all to imports. If we don't use these files other than to make this report, these should probably be unified into one script without IO between them.

File: workflows/offline_ml_diags/tests/test__ml_prediction.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/tests/test__ml_prediction.py:1]: Yes, it looks like what predict_function actually does (other than predict, which we already have as `model.predict`) is insert grid data and insert zero-values for (p)Q(1/2). I'll refactor these tests to test predict_function for these behaviors instead of completely removing them.

File: external/fv3fit/tests/training/test_cli.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_cli.py:128]: I'll probably do this as a quick separate PR before merging this in. I think we'd like to minimize the number of times the yaml API will change, so this is the time to get it in.

Comment (mcgibbon) [reply external/fv3fit/tests/training/test_cli.py:128]: OK it does actually need to be in this PR, because of how the old config treats `data_path`.

Comment (brianhenn) [reply external/fv3fit/tests/training/test_cli.py:128]: The hybrid mapper isn't actually incompatible with BatchesConfig is it? I was actually quite happy to see the vague typing of the old data path (either or a string, a list of strings, or a list of a single string) go away.

File: external/fv3fit/fv3fit/train.py
Comment (brianhenn) [reply external/fv3fit/fv3fit/train.py:17]: That sounds good to me

Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py]: Done.

Comment (brianhenn) [reply external/fv3fit/fv3fit/train.py:17]: One other docs issue: fv3fit/docs/configuration.rst is stale and should get deleted or updated to autodoc point at the right thing, since it's picked up by sphix.

File: workflows/offline_ml_diags/offline_ml_diags/compute_diags.py
Comment (brianhenn) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py:319]: OK thanks, though I'm still confused as to whether a derived model's 'derived' output will be evaluated in the offline diags. The current docs say no: http://vulcanclimatemodeling.com/docs/fv3fit/derived_model.html, but it sounds like there's been a change such that the answer is actually yes.

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py]: Done.

**Comment by mcgibbon** [comment]
I will also move the configuration classes in `fv3fit._shared.config` to the places where their training functions are defined, but I'm leaving them in-place for this PR so the changelog is readable.

File: external/fv3fit/tests/test__config.py
Comment (mcgibbon) [reply external/fv3fit/tests/test__config.py:66]: Updated, though it's really not great that RandomForestHyperparameters has to reference the TrainingConfig in this way. In the future we can add a `additional_variables` property to Hyperparameters classes which returns the variable names needed. Then this can be done programmatically, without having a hard-code at the top level for e.g. `scaler_type` implying a certain value.

File: workflows/offline_ml_diags/offline_ml_diags/compute_diags.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py]: Done. I tried looking at removing it entirely, but it's used in a way I don't understand in the `MergeOverlappingData` GeoMapper. This appears to be the only use.

File: workflows/offline_ml_diags/offline_ml_diags/compute_diags.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py:320]: Good catch, I think those docs are out of date. This should now be documented in the TrainingConfig, since you no longer need to create that model manually. I'll expand the TrainingConfig doc a little and delete that page.

This also reminds me i need to update the docs to reference the new config objects, they still reference ModelTrainingConfig.

File: workflows/offline_ml_diags/offline_ml_diags/create_report.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/offline_ml_diags/create_report.py]: Done.

**Comment by mcgibbon** [comment]
Got through a fair number of changes today, will be leaving the training refactors for tomorrow.

File: workflows/offline_ml_diags/tests/test__ml_prediction.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/tests/test__ml_prediction.py]: Done.

**Review by mcgibbon — COMMENTED** [review]
Should be ready for another review.

File: external/fv3fit/tests/training/test_cli.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_cli.py]: It's not incompatible, and the vague typing is sort of gone. The difference now is that `data_path` is a kwarg like any other, instead of being its own special thing that must exist for every batches/mapper function. If we add configuration classes similar to the Hyperparameters classes for loaders, we can also restore strict typing.

File: tests/end_to_end_integration/argo.yaml
Comment (brianhenn) [reply tests/end_to_end_integration/argo.yaml:9]: But did @oliverwm1 also review the argo workflow files, in addition to the end to end test?

Comment (oliverwm1) [reply tests/end_to_end_integration/argo.yaml:9]: yeah, the changes looked good to me.

File: workflows/offline_ml_diags/offline_ml_diags/compute_diags.py
Comment (AnnaKwa) [reply workflows/offline_ml_diags/offline_ml_diags/compute_diags.py:320]: Thanks, good catch, those docs are out of date. The 'derived' outputs will be evaluated in the offline diags now.

**Review by AnnaKwa — APPROVED** [review]
Thanks! The changes look good to me. I'm approving assuming you're getting the integration tests passing and the hangup is a config typo rather than anything major. If that's not the case let me know.

File: external/fv3fit/tests/training/test_main.py
Comment (brianhenn) [inline external/fv3fit/tests/training/test_main.py:160]: Is the parameterization of `additional_variables` actually relevant to this test?

File: external/fv3fit/tests/training/test_main.py
Comment (brianhenn) [inline external/fv3fit/tests/training/test_main.py:185]: Same comment as above re: `additional_variables`

File: external/fv3fit/tests/training/test_main.py
Comment (brianhenn) [inline external/fv3fit/tests/training/test_main.py:209]: Same comment as above re: `additional_variables`

File: external/fv3fit/tests/training/test_main.py
Comment (brianhenn) [inline external/fv3fit/tests/training/test_main.py:140]: Is the parameterization of `derived_output_variables` relevant to this test?

File: external/fv3fit/tests/training/test_main.py
Comment (brianhenn) [inline external/fv3fit/tests/training/test_main.py:303]: I don't think the case that `args.validation_data_config` is `None` is actually tested by the parameterization. Would be good to check that.

File: workflows/offline_ml_diags/tests/test__ml_prediction.py
Comment (brianhenn) [inline workflows/offline_ml_diags/tests/test__ml_prediction.py:105]: leftover?

File: tests/end_to_end_integration/argo.yaml
Comment (brianhenn) [reply tests/end_to_end_integration/argo.yaml:9]: I guess as long as the end-to-end is passing presumably the workflows are OK

**Review by brianhenn — APPROVED** [review]
Thanks @mcgibbon I think this looks good overall, I have a number of small fixes I'd ask get implemented (particularly in the fv3fit's `test_main` module).

File: external/fv3fit/docs/derived_model.rst
Comment (mcgibbon) [reply external/fv3fit/docs/derived_model.rst:3]: Basically yeah. I think you can technically still make a manual yaml after running the training, but that's harder to do and not the way we'd recommend it be done.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:145]: It could be, if load_batches is only called correctly depending on whether derived output variables are requested. If it is it's a bug, but that's why the test is here.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:176]: Similar to above, it could be, particularly if a bug makes it so derived variables are only correctly applied if additional variables are/aren't set.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py]: Removed the docstrings instead, the names here should be descriptive enough now that the tests are simpler (split out).

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:201]: It's possible for a bug to make this matter.

In each of these tests, I would maybe exclude the less-likely-to-matter cases if they took >1s to run, but all the main tests collectively take much less than a second to run because of the mocks.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:225]: Same reply as above.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py]: Fixed.

File: external/fv3fit/tests/training/test_main.py
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_main.py:331]: Fixed. This was untested in the main routines as well.

File: external/loaders/loaders/_config.py
Comment (mcgibbon) [reply external/loaders/loaders/_config.py:63]: I tried to follow the "document on the class" because there is no init routine on those subclasses. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

**Pull request merged**
